send/slack

```diff
@@ -0,0 +1,518 @@
+		elif (
+			our_user["firstName"] != their_user["profile"]["first_name"]
+			or our_user["lastName"] != their_user["profile"]["last_name"]
+			or our_user["email"] != their_user["profile"]["email"]
+			or their_user["deleted"]
+			or their_user["is_restricted"]
+			or their_user["is_ultra_restricted"]
+		):
+			update_user(client, our_user, their_user)
+
+	return deactivated_users
+
+
+def create_channel(webclient, name, is_private):
+	print_debug(f"Creating {'public' if not is_private else 'private'} channel {name}")
+	try:
+		response = webclient.conversations_create(name=name, is_private=is_private)
+	except SlackApiError as e:
+		if e.response.get("error") == "name_taken":
+			WARNING.append(
+				f"Channel '{name}' already exists, but bot not integrated. Skipping."
+			)
+			return None
+
+	check_webclient_response(response)
+	global CHANNELS_CREATED
+	CHANNELS_CREATED += 1
+	return response["channel"]
+
+
+def get_their_channels(webclient):
+	print_debug("Fetching Slack channels")
+	channels = []
+	cursor = None
+	while cursor != "":
+		response = webclient.conversations_list(
+			cursor=cursor,
+			exclude_archived=False,
+			limit=COUNT,
+			types="public_channel,private_channel",
+		)
+		check_webclient_response(response)
+		channels.extend(response["channels"])
+		cursor = response["response_metadata"].get("next_cursor")
+
+	return channels
+
+
+def get_their_users_web(webclient):
+	print_debug("Fetching Slack users")
+	users = []
+	cursor = None
+	while cursor != "":
+		response = webclient.users_list(cursor=cursor, limit=COUNT)
+		check_webclient_response(response)
+		users.extend(response["members"])
+		cursor = response["response_metadata"].get("next_cursor")
+	return users
+
+
+def get_their_members(webclient, channel_id):
+	print_debug(f"Fetching members of channel {channel_id}")
+	members = []
+	cursor = None
+	while cursor != "":
+		response = webclient.conversations_members(
+			cursor=cursor,
+			channel=channel_id,
+			limit=COUNT,
+		)
+		check_webclient_response(response)
+		members.extend(response["members"])
+		cursor = response["response_metadata"].get("next_cursor")
+
+	return members
+
+
+def invite_members(webclient, members, channel_id):
+	print_debug(f"Inviting user {members} to {channel_id}")
+	response = webclient.conversations_invite(channel=channel_id, users=members)
+	if response.get("ok", False) and response.get("error", "").startswith(
+		"already_in_channel"
+	):
+		if len(members) == 1:
+			print_debug(
+				f"Tried to invite existing member {members} of channel {channel_id}, skipped."
+			)
+			return
+		else:
+			for member in members:
+				invite_members(webclient, [member], channel_id)
+
+	check_webclient_response(response)
+	global MEMBERS_ADDED
+	MEMBERS_ADDED += len(members)
+
+
+def kick_members(webclient, members, channel_id):
+	for user in members:
+		print_debug(f"Kicking user {user} from channel {channel_id}")
+		response = webclient.conversations_kick(channel=channel_id, user=user)
```

Comment:
this could potentially be a lot of calls. Could this be a problem? (maybe the SDK does some buffering magic?) One SCIM API call should do the trick as well, no?

Reply:
i didn't find bulk operation for that

Reply:
Mixed up slack groups and channels, my bad.